test/api/test-draw.c

```diff
@@ -330,43 +343,43 @@ test_hb_draw_ttf_parser_tests (void)
     hb_font_t *font = hb_font_create (face);
     hb_face_destroy (face);
     {
-      user_data.consumed = 0;
-      g_assert (hb_font_draw_glyph (font, 0, funcs, &user_data));
+      draw_data.consumed = 0;
+      hb_font_get_glyph_shape (font, 0, funcs, &draw_data);
       char expected[] = "M50,0L50,750L450,750L450,0L50,0Z";
-      g_assert_cmpmem (str, user_data.consumed, expected, sizeof (expected) - 1);
+      g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
     }
     {
-      user_data.consumed = 0;
-      g_assert (hb_font_draw_glyph (font, 1, funcs, &user_data));
+      draw_data.consumed = 0;
+      hb_font_get_glyph_shape (font, 1, funcs, &draw_data);
       char expected[] = "M56,416L56,487L514,487L514,416L56,416ZM56,217L56,288L514,288L514,217L56,217Z";
-      g_assert_cmpmem (str, user_data.consumed, expected, sizeof (expected) - 1);
+      g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
     }
     {
-      user_data.consumed = 0;
-      g_assert (hb_font_draw_glyph (font, 4, funcs, &user_data));
+      draw_data.consumed = 0;
+      hb_font_get_glyph_shape (font, 4, funcs, &draw_data);
       char expected[] = "M332,468L197,468L197,0L109,0L109,468L15,468L15,509L109,539"
 			"L109,570Q109,674 155,720Q201,765 283,765Q315,765 342,760"
 			"Q368,754 387,747L364,678Q348,683 327,688Q306,693 284,693"
 			"Q240,693 219,664Q197,634 197,571L197,536L332,536L332,468Z"
 			"M474,737Q494,737 510,724Q525,710 525,681Q525,653 510,639"
 			"Q494,625 474,625Q452,625 437,639Q422,653 422,681"
 			"Q422,710 437,724Q452,737 474,737ZM517,536L517,0L429,0L429,536L517,536Z";
-      g_assert_cmpmem (str, user_data.consumed, expected, sizeof (expected) - 1);
+      g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
     }
     {
-      user_data.consumed = 0;
-      g_assert (hb_font_draw_glyph (font, 5, funcs, &user_data));
+      draw_data.consumed = 0;
+      hb_font_get_glyph_shape (font, 5, funcs, &draw_data);
       char expected[] = "";
-      g_assert_cmpmem (str, user_data.consumed, expected, sizeof (expected) - 1);
+      g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
```

Comment:
Allowing empty contours emit should have changed expectation of this and the next tests also like what original effect of that https://github.com/harfbuzz/harfbuzz/commit/3a98c7fae196fd149078af4553b966dc1c8d763b weird :/ I should have a closer look at it, if you can also please consider having a look at it, thanks

Reply:
Empty can have many meanings. We still don't emit a moveto until a lineto/quadritcto/curveto... I gave up on that... So I think that's why you don't see a drastic change.

Reply:
Okay debugged this a bit,

5th glyph of the test font glyphs.ttf according to fonttools is:

```
    <TTGlyph name="space" xMin="0" yMin="0" xMax="0" yMax="0">
      <contour>
        <pt x="0" y="0" on="0"/>
      </contour>
      <instructions/>
    </TTGlyph>
```

And this can make it print the empty contour apparently,

```diff
diff --git a/src/hb-ot-glyf-table.hh b/src/hb-ot-glyf-table.hh
index dee465a6b..34170ab36 100644
--- a/src/hb-ot-glyf-table.hh
+++ b/src/hb-ot-glyf-table.hh
@@ -1202,7 +1202,10 @@ struct glyf
              draw_session->move_to (font->em_fscalef_x (mid.x), font->em_fscalef_y (mid.y));
            }
            else
+           {
              first_offcurve = p;
+             last_offcurve = p;
+           }
          }
        }
        else
diff --git a/test/api/test-draw.c b/test/api/test-draw.c
index 72b1aadfb..7ac616884 100644
--- a/test/api/test-draw.c
+++ b/test/api/test-draw.c
@@ -369,7 +369,7 @@ test_hb_draw_ttf_parser_tests (void)
     {
       draw_data.consumed = 0;
       hb_font_get_glyph_shape (font, 5, funcs, &draw_data);
-      char expected[] = "";
+      char expected[] = "M0,0Q15,0 15,0L0,0Z";
       g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
     }
     {
@@ -378,7 +378,8 @@ test_hb_draw_ttf_parser_tests (void)
       char expected[] = "M346,468L211,468L211,0L123,0L123,468L29,468L29,509L123,539"
                        "L123,570Q123,674 169,720Q215,765 297,765Q329,765 356,760"
                        "Q382,754 401,747L378,678Q362,683 341,688Q320,693 298,693"
-                       "Q254,693 233,664Q211,634 211,571L211,536L346,536L346,468Z";
+                       "Q254,693 233,664Q211,634 211,571L211,536L346,536L346,468Z"
+                       "M0,0Q15,0 15,0L0,0Z";
       g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
     }
 
 ```

I'd suggest now that the expectation of the code is going to be changed it would be a good idea to revisit the logic of `consume_point` and see if it matches trusted codes you have somewhere else. Touching it probably changes expectation of edge cases tests but guess updating those worth the hassle.

Reply:
@ebraminio Curious. Where that 15 come from?!! A lone point contour in a TTF outline.... I have no idea how that is to be interpretted!

Reply:
I think it all the numbers involved should be 0.

Reply:
@ebraminio I suggest you bring it up with @RazrFalcon and we commit whatever fix he comes up with.

Reply:
My reading of the code suggests that your suggested change is correct.

Reply:
The 15 is coming from a lack of prior `move_to`... so.. to me this change is not working correctly yet :(.

Reply:
My proposed change is now:

```diff
diff --git a/src/hb-ot-glyf-table.hh b/src/hb-ot-glyf-table.hh
index dee465a6b..2c26a8df1 100644
--- a/src/hb-ot-glyf-table.hh
+++ b/src/hb-ot-glyf-table.hh
@@ -1251,6 +1251,12 @@ struct glyf
                                       font->em_fscalef_x (first_oncurve.x), font->em_fscalef_y (first_oncurve.y));
          else if (first_oncurve.has_data)
            draw_session->line_to (font->em_fscalef_x (first_oncurve.x), font->em_fscalef_y (first_oncurve.y));
+         else if (first_offcurve.has_data)
+         {
+           float x = font->em_fscalef_x (first_offcurve.x), y = font->em_fscalef_x (first_offcurve.y);
+           draw_session->move_to (x, y);
+           draw_session->quadratic_to (x, y, x, y);
+         }
 
          /* Getting ready for the next contour */
          first_oncurve = first_offcurve = last_offcurve = optional_point_t ();
diff --git a/test/api/test-draw.c b/test/api/test-draw.c
index 72b1aadfb..023e82188 100644
--- a/test/api/test-draw.c
+++ b/test/api/test-draw.c
@@ -369,7 +369,7 @@ test_hb_draw_ttf_parser_tests (void)
     {
       draw_data.consumed = 0;
       hb_font_get_glyph_shape (font, 5, funcs, &draw_data);
-      char expected[] = "";
+      char expected[] = "M15,0Q15,0 15,0Z";
       g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
     }
     {
@@ -378,7 +378,8 @@ test_hb_draw_ttf_parser_tests (void)
       char expected[] = "M346,468L211,468L211,0L123,0L123,468L29,468L29,509L123,539"
                        "L123,570Q123,674 169,720Q215,765 297,765Q329,765 356,760"
                        "Q382,754 401,747L378,678Q362,683 341,688Q320,693 298,693"
-                       "Q254,693 233,664Q211,634 211,571L211,536L346,536L346,468Z";
+                       "Q254,693 233,664Q211,634 211,571L211,536L346,536L346,468Z"
+                       "M15,0Q15,0 15,0Z";
       g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
     }
 
```

which makes it match freetype result on empty contours:

```diff
diff --git a/test/api/test-draw.c b/test/api/test-draw.c
index 72b1aadfb..2f129f48d 100644
--- a/test/api/test-draw.c
+++ b/test/api/test-draw.c
@@ -329,6 +329,8 @@ test_hb_draw_cff2 (void)
   hb_font_destroy (font);
 }
 
+#include "hb-ft.h"
+
 static void
 test_hb_draw_ttf_parser_tests (void)
 {
@@ -341,6 +343,7 @@ test_hb_draw_ttf_parser_tests (void)
   {
     hb_face_t *face = hb_test_open_font_file ("fonts/glyphs.ttf");
     hb_font_t *font = hb_font_create (face);
+    hb_ft_font_set_funcs (font);
     hb_face_destroy (face);
     {
       draw_data.consumed = 0;
@@ -358,27 +361,28 @@ test_hb_draw_ttf_parser_tests (void)
       draw_data.consumed = 0;
       hb_font_get_glyph_shape (font, 4, funcs, &draw_data);
       char expected[] = "M332,468L197,468L197,0L109,0L109,468L15,468L15,509L109,539"
-                       "L109,570Q109,674 155,720Q201,765 283,765Q315,765 342,760"
+                       "L109,570Q109,674 155,719Q201,765 283,765Q315,765 341,759"
                        "Q368,754 387,747L364,678Q348,683 327,688Q306,693 284,693"
-                       "Q240,693 219,664Q197,634 197,571L197,536L332,536L332,468Z"
-                       "M474,737Q494,737 510,724Q525,710 525,681Q525,653 510,639"
+                       "Q240,693 218,663Q197,634 197,571L197,536L332,536L332,468Z"
+                       "M474,737Q494,737 509,723Q525,710 525,681Q525,653 509,639"
                        "Q494,625 474,625Q452,625 437,639Q422,653 422,681"
-                       "Q422,710 437,724Q452,737 474,737ZM517,536L517,0L429,0L429,536L517,536Z";
+                       "Q422,710 437,723Q452,737 474,737ZM517,536L517,0L429,0L429,536L517,536Z";
       g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
     }
     {
       draw_data.consumed = 0;
       hb_font_get_glyph_shape (font, 5, funcs, &draw_data);
-      char expected[] = "";
+      char expected[] = "M15,0Q15,0 15,0Z";
       g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
     }
     {
       draw_data.consumed = 0;
       hb_font_get_glyph_shape (font, 6, funcs, &draw_data);
       char expected[] = "M346,468L211,468L211,0L123,0L123,468L29,468L29,509L123,539"
-                       "L123,570Q123,674 169,720Q215,765 297,765Q329,765 356,760"
+                       "L123,570Q123,674 169,719Q215,765 297,765Q329,765 355,759"
                        "Q382,754 401,747L378,678Q362,683 341,688Q320,693 298,693"
-                       "Q254,693 233,664Q211,634 211,571L211,536L346,536L346,468Z";
+                       "Q254,693 232,663Q211,634 211,571L211,536L346,536L346,468Z"
+                       "M15,0Q15,0 15,0Z";
       g_assert_cmpmem (str, draw_data.consumed, expected, sizeof (expected) - 1);
     }
 
```

Reply:
Thanks. Can you push that as a commit please?

Reply:
Done.